File: java-diff-utils/src/main/java/com/github/difflib/DiffUtils.java
/*
 * Copyright 2009-2017 java-diff-utils.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.difflib;

import com.github.difflib.algorithm.DiffAlgorithmFactory;
import com.github.difflib.algorithm.DiffAlgorithmI;
import com.github.difflib.algorithm.DiffAlgorithmListener;
import com.github.difflib.algorithm.myers.MyersDiff;
import com.github.difflib.patch.AbstractDelta;
import com.github.difflib.patch.Patch;
import com.github.difflib.patch.PatchFailedException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.function.BiPredicate;

/**
 * Utility class to implement the difference and patching engine.
 */
public final class DiffUtils {

    /**
     * This factory generates the DEFAULT_DIFF algorithm for all these routines.
     */
    static DiffAlgorithmFactory DEFAULT_DIFF = MyersDiff.factory();

    /**
     * Sets the default diff algorithm factory to be used by all diff routines.
     *
     * @param factory a {@link DiffAlgorithmFactory} representing the new default diff algorithm factory.
     */
    public static void withDefaultDiffAlgorithmFactory(DiffAlgorithmFactory factory) {
        DEFAULT_DIFF = factory;
    }

    /**
     * Computes the difference between two sequences of elements using the default diff algorithm.
     *
     * @param <T> a generic representing the type of the elements to be compared.
     * @param original a {@link List} representing the original sequence of elements. Must not be {@code null}.
     * @param revised a {@link List} representing the revised sequence of elements. Must not be {@code null}.
     * @param progress a {@link DiffAlgorithmListener} representing the progress listener. Can be {@code null}.
     * @return The patch describing the difference between the original and revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> original, List<T> revised, DiffAlgorithmListener progress) {
        return DiffUtils.diff(original, revised, DEFAULT_DIFF.create(), progress);
    }

    /**
     * Computes the difference between two sequences of elements using the default diff algorithm.
     *
     * @param <T> a generic representing the type of the elements to be compared.
     * @param original a {@link List} representing the original sequence of elements. Must not be {@code null}.
     * @param revised a {@link List} representing the revised sequence of elements. Must not be {@code null}.
     * @return The patch describing the difference between the original and revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> original, List<T> revised) {
        return DiffUtils.diff(original, revised, DEFAULT_DIFF.create(), null);
    }

    /**
     * Computes the difference between two sequences of elements using the default diff algorithm.
     *
     * @param <T> a generic representing the type of the elements to be compared.
     * @param original a {@link List} representing the original sequence of elements. Must not be {@code null}.
     * @param revised a {@link List} representing the revised sequence of elements. Must not be {@code null}.
     * @param includeEqualParts a {@link boolean} representing whether to include equal parts in the resulting patch.
     * @return The patch describing the difference between the original and revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> original, List<T> revised, boolean includeEqualParts) {
        return DiffUtils.diff(original, revised, DEFAULT_DIFF.create(), null, includeEqualParts);
    }

    /**
     * Computes the difference between two strings using the default diff algorithm.
     *
     * @param sourceText a {@link String} representing the original string. Must not be {@code null}.
     * @param targetText a {@link String} representing the revised string. Must not be {@code null}.
     * @param progress a {@link DiffAlgorithmListener} representing the progress listener. Can be {@code null}.
     * @return The patch describing the difference between the original and revised strings. Never {@code null}.
     */
    public static Patch<String> diff(String sourceText, String targetText,
                                     DiffAlgorithmListener progress) {
        return DiffUtils.diff(
                Arrays.asList(sourceText.split("\n")),
                Arrays.asList(targetText.split("\n")), progress);
    }

    /**
     * Computes the difference between the original and revised list of elements
     * with default diff algorithm
     *
     * @param source a {@link List} representing the original text. Must not be {@code null}.
     * @param target a {@link List} representing the revised text. Must not be {@code null}.
     * @param equalizer a {@link BiPredicate} representing the equalizer object to replace the default compare
     * algorithm (Object.equals). If {@code null} the default equalizer of the
     * default algorithm is used.
     * @return The patch describing the difference between the original and
     * revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> source, List<T> target,
                                    BiPredicate<T, T> equalizer) {
        if (equalizer != null) {
            return DiffUtils.diff(source, target,
                    DEFAULT_DIFF.create(equalizer));
        }
        return DiffUtils.diff(source, target, new MyersDiff<>());
    }

    public static <T> Patch<T> diff(List<T> original, List<T> revised,
                                    DiffAlgorithmI<T> algorithm, DiffAlgorithmListener progress) {
        return diff(original, revised, algorithm, progress, false);
    }

    /**
     * Computes the difference between the original and revised list of elements
     * with default diff algorithm
     *
     * @param original a {@link List} representing the original text. Must not be {@code null}.
     * @param revised a {@link List} representing the revised text. Must not be {@code null}.
     * @param algorithm a {@link DiffAlgorithmI} representing the diff algorithm. Must not be {@code null}.
     * @param progress a {@link DiffAlgorithmListener} representing the diff algorithm listener.
     * @param includeEqualParts Include equal data parts into the patch.
     * @return The patch describing the difference between the original and
     * revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> original, List<T> revised,
                                    DiffAlgorithmI<T> algorithm, DiffAlgorithmListener progress,
                                    boolean includeEqualParts) {
        Objects.requireNonNull(original, "original must not be null");
        Objects.requireNonNull(revised, "revised must not be null");
        Objects.requireNonNull(algorithm, "algorithm must not be null");

        return Patch.generate(original, revised, algorithm.computeDiff(original, revised, progress), includeEqualParts);
    }


    /**
     * Computes the difference between the original and revised list of elements
     * with default diff algorithm
     *
     * @param original a {@link List} representing the original text. Must not be {@code null}.
     * @param revised a {@link List} representing the revised text. Must not be {@code null}.
     * @param algorithm a {@link DiffAlgorithmI} representing the diff algorithm. Must not be {@code null}.
     * @return The patch describing the difference between the original and
     * revised sequences. Never {@code null}.
     */
    public static <T> Patch<T> diff(List<T> original, List<T> revised, DiffAlgorithmI<T> algorithm) {
        return diff(original, revised, algorithm, null);
    }

    /**
     * Computes the difference between the given texts inline. This one uses the
     * "trick" to make out of texts lists of characters, like DiffRowGenerator
     * does and merges those changes at the end together again.
     *
     * @param original a {@link String} representing the original text. Must not be {@code null}.
     * @param revised a {@link String} representing the revised text. Must not be {@code null}.
     * @return The patch describing the difference between the original and
     * revised sequences. Never {@code null}.
     */
    public static Patch<String> diffInline(String original, String revised) {
        List<String> origList = new ArrayList<>();
        List<String> revList = new ArrayList<>();
        for (Character character : original.toCharArray()) {
            origList.add(character.toString());
        }
        for (Character character : revised.toCharArray()) {
            revList.add(character.toString());
        }
        Patch<String> patch = DiffUtils.diff(origList, revList);
        for (AbstractDelta<String> delta : patch.getDeltas()) {
            delta.getSource().setLines(compressLines(delta.getSource().getLines(), ""));
            delta.getTarget().setLines(compressLines(delta.getTarget().getLines(), ""));
        }
        return patch;
    }

    /**
     * Applies the given patch to the original list and returns the revised list.
     *
     * @param original a {@link List} representing the original list.
     * @param patch a {@link List} representing the patch to apply.
     * @return the revised list.
     * @throws PatchFailedException if the patch cannot be applied.
     */
    public static <T> List<T> patch(List<T> original, Patch<T> patch)
            throws PatchFailedException {
        return patch.applyTo(original);
    }

    /**
     * Applies the given patch to the revised list and returns the original list.
     *
     * @param revised a {@link List} representing the revised list.
     * @param patch a {@link Patch} representing the patch to apply.
     * @return the original list.
     * @throws PatchFailedException if the patch cannot be applied.
     */
    public static <T> List<T> unpatch(List<T> revised, Patch<T> patch) {
        return patch.restore(revised);
    }

    private static List<String> compressLines(List<String> lines, String delimiter) {
        if (lines.isEmpty()) {
            return Collections.emptyList();
        }
        return Collections.singletonList(String.join(delimiter, lines));
    }

    private DiffUtils() {
    }
}

File: src/commonMain/kotlin/com/github/difflib/DiffUtils.kt
/*
 * Copyright 2009-2017 java-diff-utils.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.difflib

import com.github.difflib.algorithm.DiffAlgorithmI
import com.github.difflib.algorithm.DiffAlgorithmListener
import com.github.difflib.algorithm.DiffException
import com.github.difflib.algorithm.myers.MyersDiff
import com.github.difflib.patch.AbstractDelta
import com.github.difflib.patch.Patch
import com.github.difflib.patch.PatchFailedException

internal typealias Predicate<T> = (T) -> Boolean
internal typealias BiPredicate<T,R> = (T, R) -> Boolean
internal typealias Consumer<T> = (T) -> Unit
internal typealias Function<T,R> = (T) -> R

/**
 * Implements the difference and patching engine
 *
 * @author [Dmitry Naumenko](dm.naumenko@gmail.com)
 */
object DiffUtils {

    /**
     * Computes the difference between the original and revised list of elements with default diff algorithm
     *
     * @param original The original text. Must not be `null`.
     * @param revised The revised text. Must not be `null`.
     * @param progress progress listener
     * @return The patch describing the difference between the original and revised sequences. Never `null`.
     * @throws com.github.difflib.algorithm.DiffException
     */
//    @Throws(DiffException::class)
    fun <T> diff(original: List<T>, revised: List<T>, progress: DiffAlgorithmListener): Patch<T> {
        return DiffUtils.diff(original, revised, MyersDiff(), progress)
    }

//    @Throws(DiffException::class)
    fun <T> diff(original: List<T>, revised: List<T>): Patch<T> {
        return DiffUtils.diff(original, revised, MyersDiff(), null)
    }

    /**
     * Computes the difference between the original and revised text.
     */
//    @Throws(DiffException::class)
    fun diff(sourceText: String, targetText: String,
             progress: DiffAlgorithmListener): Patch<String> {
        return DiffUtils.diff(
                sourceText.split("\n".toRegex()).dropLastWhile { it.isEmpty() },
                targetText.split("\n".toRegex()).dropLastWhile { it.isEmpty() }, progress)
    }

    /**
     * Computes the difference between the original and revised list of elements with default diff algorithm
     *
     * @param source The original text. Must not be `null`.
     * @param target The revised text. Must not be `null`.
     *
     * @param equalizer the equalizer object to replace the default compare algorithm (Object.equals). If `null`
     * the default equalizer of the default algorithm is used..
     * @return The patch describing the difference between the original and revised sequences. Never `null`.
     */
//    @Throws(DiffException::class)
    fun <T> diff(source: List<T>, target: List<T>,
                 equalizer: BiPredicate<T, T>?): Patch<T> {
        return if (equalizer != null) {
            DiffUtils.diff(source, target,
                    MyersDiff(equalizer))
        } else DiffUtils.diff(source, target, MyersDiff())
    }

    /**
     * Computes the difference between the original and revised list of elements with default diff algorithm
     *
     * @param original The original text. Must not be `null`.
     * @param revised The revised text. Must not be `null`.
     * @param algorithm The diff algorithm. Must not be `null`.
     * @param progress The diff algorithm listener.
     * @return The patch describing the difference between the original and revised sequences. Never `null`.
     */
//    @Throws(DiffException::class)
    fun <T> diff(original: List<T>, revised: List<T>,
                 algorithm: DiffAlgorithmI<T>, progress: DiffAlgorithmListener?): Patch<T> {
        return Patch.generate(original, revised, algorithm.computeDiff(original, revised, progress))
    }

    /**
     * Computes the difference between the original and revised list of elements with default diff algorithm
     *
     * @param original The original text. Must not be `null`.
     * @param revised The revised text. Must not be `null`.
     * @param algorithm The diff algorithm. Must not be `null`.
     * @return The patch describing the difference between the original and revised sequences. Never `null`.
     */
//    @Throws(DiffException::class)
    fun <T> diff(original: List<T>, revised: List<T>,
                 algorithm: DiffAlgorithmI<T>): Patch<T> {
        return diff(original, revised, algorithm, null)
    }

    /**
     * Computes the difference between the given texts inline. This one uses the "trick" to make out of texts lists of
     * characters, like DiffRowGenerator does and merges those changes at the end together again.
     *
     * @param original
     * @param revised
     * @return
     */
//    @Throws(DiffException::class)
    fun diffInline(original: String, revised: String): Patch<String> {
        val origList = ArrayList<String>()
        val revList = ArrayList<String>()
        for (character in original) {
            origList.add(character.toString())
        }
        for (character in revised) {
            revList.add(character.toString())
        }
        val patch = DiffUtils.diff(origList, revList)
        for (delta in patch.deltas) {
            delta.source.lines = compressLines(delta.source.lines!!, "")
            delta.target.lines = compressLines(delta.target.lines!!, "")
        }
        return patch
    }

    private fun compressLines(lines: List<String>, delimiter: String): List<String> {
        return if (lines.isEmpty()) {
            emptyList<String>()
        } else listOf<String>(lines.joinToString(delimiter))
    }

    /**
     * Patch the original text with given patch
     *
     * @param original the original text
     * @param patch the given patch
     * @return the revised text
     * @throws PatchFailedException if can't apply patch
     */
//    @Throws(PatchFailedException::class)
    fun <T> patch(original: List<T>, patch: Patch<T>): List<T> {
        return patch.applyTo(original)
    }

    /**
     * Unpatch the revised text for a given patch
     *
     * @param revised the revised text
     * @param patch the given patch
     * @return the original text
     */
    fun <T> unpatch(revised: List<T>, patch: Patch<T>): List<T> {
        return patch.restore(revised)
    }
}
